src/theory/quantifiers/sygus/cegis_core_connective.cpp

```diff
@@ -0,0 +1,843 @@
+    }
+  } while (!mvId.isNull() && addSuccess);
+  if (!addSuccess)
+  {
+    Trace("sygus-ccore") << ">>> Failed to generate initial candidate"
+                         << std::endl;
+    return Node::null();
+  }
+  Trace("sygus-ccore") << "----- Initial candidate is " << an << std::endl;
+
+  do
+  {
+    addSuccess = false;
+    // try a new core
+    SmtEngine checkSol(nm->toExprManager());
+    checkSol.setIsInternalSubsolver();
+    checkSol.setLogic(smt::currentSmtEngine()->getLogicInfo());
+    Trace("sygus-ccore") << "----- Check candidate " << an << std::endl;
+    std::vector<Node> rasserts = asserts;
+    rasserts.push_back(ccheck.getFormula());
+    std::shuffle(rasserts.begin(), rasserts.end(), Random::getRandom());
+    Node query = rasserts.size() == 1 ? rasserts[0] : nm->mkNode(AND, rasserts);
+    for (const Node& a : rasserts)
+    {
+      checkSol.assertFormula(a.toExpr());
+    }
+    Result r = checkSol.checkSat();
+    Trace("sygus-ccore") << "----- check-sat returned " << r << std::endl;
+    if (r.asSatisfiabilityResult().isSat() == Result::UNSAT)
+    {
+      // it entails the postcondition
+      // get the unsat core
+      std::vector<Node> uasserts;
+      bool hasQuery = getUnsatCore(checkSol, ccheck.getFormula(), uasserts);
+      // now, check the side condition
+      bool falseCore = false;
+      if (!d_sc.isNull())
+      {
+        if (!hasQuery)
+        {
+          // already know false
+          falseCore = true;
+        }
+        else
+        {
+          Trace("sygus-ccore") << "----- Check side condition" << std::endl;
+          SmtEngine checkSc(nm->toExprManager());
+          checkSc.setIsInternalSubsolver();
+          checkSc.setLogic(smt::currentSmtEngine()->getLogicInfo());
+          std::vector<Node> scasserts;
+          scasserts.insert(scasserts.end(), uasserts.begin(), uasserts.end());
+          scasserts.push_back(d_sc);
+          std::shuffle(scasserts.begin(), scasserts.end(), Random::getRandom());
+          Result rsc = checkSc.checkSat();
+          Trace("sygus-ccore")
+              << "----- check-sat returned " << rsc << std::endl;
+          if (rsc.asSatisfiabilityResult().isSat() == Result::UNSAT)
+          {
+            // can minimize based on this
+            uasserts.clear();
+            getUnsatCore(checkSc, d_sc, uasserts);
+            falseCore = true;
+          }
+        }
+      }
+
+      if (!falseCore)
+      {
+        Trace("sygus-ccore") << ">>> Solution : " << uasserts << std::endl;
+
+        Node sol = ccheck.getSygusSolution(uasserts);
+        Trace("sygus-ccore-sy") << "Sygus solution : " << sol << std::endl;
+        return sol;
+      }
+      else
+      {
+        Assert(!uasserts.empty());
+        Node xu = uasserts[0];
+        Trace("sygus-ccore")
+            << "--- Add false core : " << uasserts << std::endl;
+        if (uasserts.size() == 1)
+        {
+          // singleton false core should be removed from pool TODO
```

Comment:
Is this intentional or was overlooked?

Reply:
I've added this comment as a potential for future optimization, will clean this.